### ISSUE TITLE
Increase paralellism of importers

### ISSUE THREAD

**Issue by orangejulius** [pelias/docker#72]
We [currently](https://github.com/pelias/docker/blob/master/cmd/import.sh#L24-L34) run Pelias importers in series in this repository.

This keeps memory and CPU requirements low, and was initially required to avoid fatal errors if Elasticsearch became overloaded.

However, over time we've [made improvements](https://github.com/pelias/docker/commit/0a16d385441dab1126373b77f6fbfafe47598f4f) and are now pretty confident that as long as all the importers can fit into memory, an import can finish just fine even on a very CPU constrained machine.

Running importers in parallel would allow imports to finish much faster where more CPU is available, so we should consider making that at least an option, if not the default, in this project.

**Comment by orangejulius** [comment]
Hmm, on the other hand, defaulting to running multiple importers in parallel will mean the memory requirements are significant for larger builds. We might have to keep sequential importing the default.

**Comment by missinglink** [comment]
It's a good idea, I think it's important to keep the barriers of entry low for the docker-compose projects.

They are a really easy way to get Pelias up-and-running for testing and evaluation and I'd prefer not to sacrifice that simplicity if we can avoid it.

The community really likes the docker-compose setup and we've now got projects for Europe and the planet which are surfacing problems such as memory requirements and slow build times which weren't really considered when this repo was conceived.

I'm still on the fence as to whether we should encourage use of docker-compose for building large-scale production-ready clusters (even though it's been proven that it's possible to do so).

@orangejulius what are your general thoughts on where you see this repo going? Do we say it's just for dev work and getting started or do we go all-in and say it's 'the way to build pelias'?

**Comment by orangejulius** [comment]
Okay, after thinking about this more, I think parallelism as a default in this repo would be a mistake:

- it would increase the hardware requirements of a build
- it would increase the complexity and chance of failure of a build

I think the strategy for this project should be as follows:
- make it as easy as possible to get set up quickly with Pelias
- but, where possible, provide primitive building blocks to construct more advanced functionality

This would mean we keep things simple by default, and don't support things like parallelism, running behind a proxy, etc.

**Comment by bboure** [comment]
@orangejulius @missinglink 

I have been using this project for local/dev/testing purposes on a small scale (cities) and it works great.
I am now looking into how I will eventually go to production, involving a planet build. 

What would be the recommended way of running a production build then?
The documentation is pretty fuzzy about that.
The [Kubernetes](https://github.com/pelias/kubernetes/tree/master/build) repo says:

> However, you probably shouldn't use it, as it hasn't been maintained in some time. Instead, use the Pelias Docker setup to run a build, and save the Elasticsearch index and other data in locations where they can be used by your primary Pelias installation.

But here you seem to say the opposite.

> I'm still on the fence as to whether we should encourage use of docker-compose for building large-scale production-ready clusters (even though it's been proven that it's possible to do so).

Well, OK, this comment is quite old. Has something changed meanwhile?

How do you guys do it at https://geocode.earth/ ?

Thanks

**Comment by orangejulius** [comment]
Hi @bboure,
Ah, this is the trouble with old conversations being recorded on GitHub. It's easy for us to inadvertently send mixed messages :(

I'm sure we said it somewhere, but the Docker setup in this repository **is** the recommended way to do all Pelias builds, including planet size, and it's what we use at Geocode Earth.

You might want to run some importers in parallel on a big enough machine with a planet build. As discussed in this issue, we don't do that out of the box to ensure that people trying out Pelias on a small machine have the best chance of a good experience, but it's not hard to do.

Something as simple as
```
pelias import oa &
pelias import osm &
pelias import wof &
...etc
wait
```

in a bash script will work as a starting point.

**Comment by bboure** [comment]
Great, thanks for the clarification @orangejulius 


In AWS (which I believe yu use), what would you use? a simple EC2 instance with docker installed or do you have some more advanced setup with maybe ECS?

Do you have automated processes or is it rather a more manual thing?
(like ssh into the machine and run the script).

It would help a lot if you could share your experience on this.

Thanks again

**Comment by NickStallman** [comment]
Bare metal is also doable which is what I do, although I'll probably migrate to Docker at some point. :)

Using AWS can get expensive over time so I do builds on a Threadripper 1900X with 128gigs of RAM and 2x NVMe SSDs.

My live production server is somewhat smaller as Pelias doesn't need as much horsepower or RAM for good query performance.

**Comment by orangejulius** [comment]
We run our builds on regular `c5d.9xlarge` instances with a simple AMI to set up some basic tooling. The NVMe and high core count is key. Any sort of cluster is just wasteful for builds, since they only run for a few hours and do not require redundancy (losing a build is not an emergency).

Back at Mapzen we (foolshly) used a cluster of 4 smaller Elasticsearch instances and it was massively more complicated and slower.

As long as Pelias is in the realm of "biggish" data (won't perform well on a personal computer but does not need more hardware than can be provided by a single easily acquired machine from a cloud or VPS provider), anything like Kubernetes, ECS, etc is overkill for builds.

Running clusters for querying is another story since there are lots of requirements around reliability, failure modes, throughput, latency, ease of maintenace over time, etc :)

**Comment by orangejulius** [comment]
We hope to write some blog posts about all this sort of stuff soon, so stay tuned!
